src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
+        uint maxDebt = (availableCollateral*10**decimals / get_daiETH_Price());
+        return maxDebt;
+    }
+
+    ///@notice Users repay their debt; in debt asset terms
+    ///@dev This covers partial and full repayment scenarioes
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
+        debts[msg.sender] -= debtAmount;
+
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
+        
+        if (debts[msg.sender] > 0){
+            uint collateralRequired = getCollateralRequired(debts[msg.sender]);
+            uint spareDeposit = deposits[msg.sender] - collateralRequired;
+            require(collateralAmount < spareDeposit, "Collateral unavailable!");
+        }
+        
+        deposits[msg.sender] -= collateralAmount;
+        bool sent = collateral.transfer(msg.sender, collateralAmount);
+        require(sent, "Withdraw failed!");
+        emit Withdraw(address(collateral), msg.sender, collateralAmount);            
+    }
+
+
+    ///@dev Calculates collateral required to support existing debt position at current market prices 
+    ///@param debtAmount Amount of debt asset to support
+    function getCollateralRequired(uint debtAmount) public view returns(uint) {
+        uint baseCollateralRequired = debtAmount*get_daiETH_Price() / (10**decimals);
+        uint bufferCollateralRequired = (baseCollateralRequired/bufferNumerator)*bufferDenominator;
+        return bufferCollateralRequired;
+    }
+
+    ///@dev Can only be called by Vault owner; triggers liquidation check on supplied user address
+    ///@param user Address of user to trigger liquidation check
+    function liquidation(address user) public onlyOwner { 
+        uint collateralRequired = getCollateralRequired(debts[user]);
+
+        if (collateralRequired > deposits[user]){
+            emit Liquidation(address(collateral), address(debt), user, debts[user], deposits[user]); 
+            delete deposits[user];
+            delete debts[user];
+        }
+    }
+
+    ///@notice Price is returned as integer extending over its decimal places
+    ///@dev Get current market price of Collateral/Debt asset from Chainlink
+    function get_daiETH_Price() public view returns(uint) {
```

Comment:
why are we using daiEth price?? im confused.  i thought this vault can take any 2 erc 20 tokens....

Reply:
Vault.sol is specific to the DAI/ETH example in the issue, while VaultV2.sol is the general implementation where one could have tokens of differing decimal precision.

Although tbh, I did not change the name of the price function in V2 - so that's definitely and error on my part.

Reply:
So sorry i just read the requirements again.  It was different than what i was remembering.  I will re review and cleanup my review tomorrow. I think v2 must have the math im looking for

Reply:
Don't worry, to make it easier for both Alberto and you, I'm going to flatten both versions into a single generic version and upload as a new branch(addressing the comments provided so far). 

Cleaner and less confusion that way :)